include/swift/AST/RewriteSystem.h

```diff
@@ -0,0 +1,411 @@
+    Protocol,
+
+    /// When appearring at the end of a term, represents that the
+    /// term conforms to the layout.
+    Layout
+  };
+
+  Kind getKind() const {
+    if (!Value) {
+      assert(Protos.size() == 1);
+      return Kind::Protocol;
+    }
+
+    if (Value.is<Identifier>()) {
+      if (!Protos.empty())
+        return Kind::AssociatedType;
+      return Kind::Name;
+    }
+
+    if (Value.is<GenericTypeParamType *>()) {
+      assert(Protos.empty());
+      return Kind::GenericParam;
+    }
+
+    if (Value.is<LayoutConstraint>()) {
+      assert(Protos.empty());
+      return Kind::Layout;
+    }
+
+    llvm_unreachable("Bad term rewriting atom");
+  }
+
+  /// Get the identifier associated with an unbound name atom or an
+  /// associated type atom.
+  Identifier getName() const {
+    assert(getKind() == Kind::Name ||
+           getKind() == Kind::AssociatedType);
+    return Value.get<Identifier>();
+  }
+
+  /// Get the single protocol declaration associate with a protocol atom.
+  const ProtocolDecl *getProtocol() const {
+    assert(getKind() == Kind::Protocol);
+    assert(Protos.size() == 1);
+    return Protos.front();
+  }
+
+  /// Get the list of protocols associated with a protocol or associated
+  /// type atom.
+  llvm::TinyPtrVector<const ProtocolDecl *> getProtocols() const {
+    assert(getKind() == Kind::Protocol ||
+           getKind() == Kind::AssociatedType);
+    assert(!Protos.empty());
+    return Protos;
+  }
+
+  /// Get the generic parameter associated with a generic parameter atom.
+  GenericTypeParamType *getGenericParam() const {
+    assert(getKind() == Kind::GenericParam);
+    return Value.get<GenericTypeParamType *>();
+  }
+
+  /// Get the layout constraint associated with a layout constraint atom.
+  LayoutConstraint getLayoutConstraint() const {
+    assert(getKind() == Kind::Layout);
+    return Value.get<LayoutConstraint>();
+  }
+
+  int compare(Atom other, const ProtocolGraph &protos) const;
+
+  void dump(llvm::raw_ostream &out) const;
+
+  friend bool operator==(Atom lhs, Atom rhs) {
+    return (lhs.Protos.size() == rhs.Protos.size() &&
+            std::equal(lhs.Protos.begin(), lhs.Protos.end(),
+                       rhs.Protos.begin()) &&
+            lhs.Value == rhs.Value);
+  }
+
+  friend bool operator!=(Atom lhs, Atom rhs) {
+    return !(lhs == rhs);
+  }
+};
+
+/// A term is a sequence of one or more atoms.
+///
+/// The first atom in the term must be a protocol, generic parameter, or
+/// associated type atom.
+///
+/// A layout constraint atom must only appear at the end of a term.
+///
+/// Out-of-line methods are documented in RewriteSystem.cpp.
+class Term final {
+  llvm::SmallVector<Atom, 3> Atoms;
+
+public:
+  Term() {}
+
+  explicit Term(llvm::SmallVector<Atom, 3> &&atoms)
+    : Atoms(atoms) {}
+
+  explicit Term(ArrayRef<Atom> atoms)
+    : Atoms(atoms.begin(), atoms.end()) {}
+
+  void add(Atom atom) {
+    Atoms.push_back(atom);
```

Comment:
Same here: emplace_back and a move are probably slight wins.

Reply:
Atom will soon become uniqued/immutable so I'll leave this alone for now.